CONTRIBUTING.md

```diff
@@ -509,7 +510,11 @@ To get a sense for what changes are considered mass rebuilds, see [previously me
 ## Commit conventions
 [commit-conventions]: #commit-conventions
 
-- Create a commit for each logical unit.
+- Create a commit for each logical unit at the Nixpkgs level.
+  This means that you should rarely break multiple changes in the same file into different logical units, because that's too fine-grained at the Nixpkgs level.
+  Some cases where you'd do so are when each change is linked to a different change in Nixpkgs as a whole.
+  For example, if you're refactoring multiple files due to the same change, this is considered a logical unit, and should be in the same commit.
+  However, fixing different packages due to unrelated issues should ideally be in different commits, even if they're included in the same PR.
```

Comment:
This might just be me, but this wording blurs my understanding of a logical unit. Previously, I would've thought that a logical unit meant a single change with a single intent, and that change or intent would be representable as a single short sentence in the commit message.

This however, seems to claim that any extra changes to the package would also be considered part of one commit. So if you have a pull request with the following commits, currently split up into what I'd previously call logical units:

```
package: 1.0.0 -> 1.1.0
         (also changes some dependencies that are a direct result
          of the package being upgraded)
package: clean up expression
package: add <name> as maintainer
package: add version test
```

Then that should be reduced to just `package: 1.0.0 -> 1.1.0`, and would still be considered a logical unit?

In my opinion, this makes the git history much more unclear and hard to read, and almost defeats the purpose of having this convention in the first place. At least, I'd not name it a logical unit, but rather a "per-package commit" or something of that sort.

Reply:
That being said, I do realise that it can be unnecessarily painstaking for both the PR author and the reviewer to go forth and back about splitting commits and ensuring everything is ordered in what I'd previously call logical units, despite all the changes being otherwise mergeable.

Reply:
Thank you for starting this discussion, since I could definitely use the help making that clearer.

It might make more sense to reword that item with the outcome we're trying to create: not having superfluous commits in the PR, to avoid overhead when reviewing commits separately and/or having to squash all commits when merging the PR.

The same conventions later mention the following:

> If you have commits pkg-name: oh, forgot to insert whitespace: squash commits in this case. Use git rebase -i.

Based on your example, I could see cases where `package: clean up expression` should remain as a single commit or all 3 bottom commits should also become a single commit.

There are too many combinations of what a committer/merger thinks should be merged with a squash commit or not, and what the PR author thinks is a logical unit, and what reviewers think is a logical unit and/or want to review.
After thinking more about this, I think it's impossible to clarify all of these cases directly, but it might still be possible to ask something like "think about whether you're creating reviewing overhead with too many commits, or whether a committer might be forced to squash your PR when merging, and try to avoid those cases".

Reply:
IMO mega-commits (except `treewide`) should be avoided. Large changes should be split into multiple commits in a way that first and foremost minimize the review effort. Each change can then also be reverted in isolation, and carry more specific and useful context in their commit message body. (The commit messages by `nix-update --commit` are a good example of why bump commits should be separate)

Example where the diff in the `/files` endpoint completely fails to show the true change compared to reviewing each commit: #271476

What we want to discourage however are fixup commits like `package: address review comments` _(unless the review is understood to take a long time and a squash will no matter what have to be a part of the final touch-ups anyway)_ or `package: do x` followed by a `package: undo x and do y` which got added after review. Those should be squashed

> fixing different packages due to unrelated issues should ideally be in different commits

This one is divisive. Some prefer separate commits per package (staging and python-updates), while others prefer all to be squashed into a single `treewide` commit because otherwise the github UI becomes noisy and unwieldy

Reply:
> There are too many combinations of what a committer/merger thinks should be merged with a squash commit or not

I think most effective way to reduce back-and-forth between reviewers and contributors would be to give contributors room for "ideally one, but up to 3 commits per package/module/unit per PR", before a reviewer have grounds to request a squash. This would allow a commit for both the intended change, a bump and a cleanup commit. "3" is subject to bikeshedding

Reply:
> What we want to discourage are fixup commits like package: address review comments

Actually I'd like to contest even that. I've done PRs (e.g. https://github.com/NixOS/nixpkgs/pull/278805) with very carefully crafted commit histories, where each commit is a logical unit. After the review, it would be very tedious to amend each commit with its own changes, potentially requiring conflict resolution at every step. So just adding a commit that addresses the review altogether saves a lot of time and sanity.

It might be best for this section to not give any concrete guidelines, but rather just list good and bad examples.